File: Hash Table/245_ShortestWordDistanceIII.java
public class Solution {
    public int shortestWordDistance(String[] words, String word1, String word2) {
        int idx1 = -1, idx2 = -1, shortest = Integer.MAX_VALUE;
        for (int i = 0; i < words.length; i++) {
            if (words[i].equals(word1)) {
                if (word1.equals(word2)) {
                    if (idx1 <= idx2) idx1 = i;
                    else idx2 = i;
                } else {
                    idx1 = i;
                }
            } else if (words[i].equals(word2)) {
                idx2 = i;
            }
            if (idx1 != -1 && idx2 != -1) shortest = Math.min(shortest, Math.abs(idx1 - idx2));
        }
        return shortest;
    }
}

File: 245_ShortestWordDistanceIII.py
class Solution(object):
    def shortestWordDistance(self, words, word1, word2):
        """
        :type words: List[str]
        :type word1: str
        :type word2: str
        :rtype: int
        """
        pos1, pos2, dis = -1, -1, sys.maxint
        for i in range(len(words)):
            curr = words[i]
            if curr == word1:
                if word1 == word2:
                    if pos1 == 1 or pos1 < pos2:
                        pos1 = i
                    else:
                        pos2 = i
                else:
                    pos1 = i
            elif curr == word2:
                pos2 = i

            if pos1 != -1 and pos2 != -1:
                dis = min(dis, abs(pos1 - pos2))
        return dis